transforms/universal/web2parquet/dpk_web2parquet/transform.py

```diff
@@ -0,0 +1,137 @@
+
+        self.count = 0
+        self.docs = []
+
+    def on_download(self, url: str, body: bytes, headers: dict) -> None:
+        """
+        Callback function called when a page has been downloaded.
+        You have access to the request URL, response body and headers.
+        """
+        doc={}
+        doc['url'] = url
+#        doc['file_size'] = int(headers.get('Content-Length', 0))  # Default to 0 if not found  
+        doc['content_type']=headers.get('Content-Type')
+        try:
+            filename = headers.get('Content-Disposition').split('filename=')[1].strip().strip('"')
+        except:
+            url_split=url.split('/')
+            filename = url_split[-1] if not url.endswith('/') else url_split[-2]
+            filename = filename.replace('.','_')+"-"+doc['content_type'].split(';')[0].replace("/", ".")
+        doc['filename']=filename
+        doc['contents'] = body
+        
+        logger.debug(f"url: {doc['url']}, filename: {doc['filename']}, content_type: {doc['content_type']}")
+
+        ## Enforce download limits
+        if len(self.docs) < self.downloads:
+            self.docs.append(doc)
+
+
+    def transform(self, table: pa.Table=None, file_name: str = None) -> tuple[list[pa.Table], dict[str, Any]]:
+        """
+        Put Transform-specific to convert one Table to 0 or more tables. It also returns
+        a dictionary of execution statistics - arbitrary dictionary
+        This implementation makes no modifications so effectively implements a copy of the
+        input parquet to the output folder, without modification.
+        """
+        start_time = time.time()
+        crawl(
+            self.seed_urls,
+            self.on_download,
+            user_agent=user_agent,
+            depth_limit=self.depth,
+            download_limit=self.downloads,
+            allow_mime_types=self.allow_mime_types
+        )  # blocking call
+
+        # Shutdown all crawls
+        # Check with @Matsubara-san as this is preventing us from calling the transfrom method a second time.
+    #    shutdown()
+
+        end_time = time.time()      
+#        logger.debug(f"Way After: {self.docs}")
+        table = pa.Table.from_pylist(self.docs)
+        metadata = {
+            "count": len(self.docs),
+            "requested_seeds": len(self.seed_urls),
+            "requested_depth": self.depth,
+            "requested_downloads": self. downloads,
+            }
+        logger.info(f"Crawling is completed in {end_time - start_time:.2f} seconds") 
+        logger.info(f"{metadata = }")
+
+        #############################################################################
+        ## The same transform can also be used to store crawled files to local folder
+        if self.folder:
+            dao=DataAccessLocal(local_config={'output_folder':self.folder,'input_folder':'.'})
+            for x in self.docs:
+                dao.save_file(self.folder+'/'+x['filename'], x['contents'])
```

Comment:
Q: How does this DAO handle a file which has the same filename as one saved in the folder before? e.g. We crawl test.com and get `test.com/path1/doc.pdf` and `test.com/path2/doc.pdf`. It looks the two have the same `filename` value.

Reply:
@hmtbr I think we are using the full URL path to form the filename (i.e. the filename for test.com/path1/doc.pdf  becomes path1_doc.pdf and test.com/path2/doc.pdf becomes path2_doc.pdf). This filename generation is not  handled by the DAO but by the transform itself.